File: apps/web/src/domain/senderClient.ts
/**
 * Sender-client origin disclosure for encrypted message payloads.
 *
 * The tag travels only inside the authenticated, encrypted inner payload, so
 * the relay never observes it and cannot forge or strip it. It is a claim
 * made by the sending client build, not an attestation: receivers surface it
 * as advisory context and fail closed when it is missing or unknown.
 */
export type SenderClient = "android" | "web";

/** The platform of this client build. Every outbound payload is tagged with it. */
export const LOCAL_SENDER_CLIENT: SenderClient = "web";

const WIRE_FIELD = "sender_client";

export function senderClientWireField(): string {
  return WIRE_FIELD;
}

/** Strict allowlist parse. Returns null for missing, mistyped, or unknown values. */
export function parseSenderClient(value: unknown): SenderClient | null {
  return value === "android" || value === "web" ? value : null;
}

export function isWebSender(client: SenderClient | undefined | null): boolean {
  return client === "web";
}

/**
 * Advisory text rendered next to a received message. Own messages never need
 * one; callers must not invoke this for locally composed messages.
 */
export function senderOriginNotice(client: SenderClient): string {
  return client === "web"
    ? "Sent from the web client: that device may lack screenshot protection, and its browser cannot guarantee memory wiping."
    : "Sent from the Android app: the sending device enforces screen-capture and memory protections.";
}

File: android/app/src/main/java/com/abyssal/chat/domain/model/SenderClient.kt
package com.abyssal.chat.domain.model

/**
 * Sender-client origin disclosure carried inside authenticated, encrypted
 * message payloads. The relay only ever sees ciphertext, so it cannot forge,
 * strip, or observe this tag. It is a claim asserted by the sending client
 * build rather than an attestation: receivers treat it as advisory context
 * and fail closed when it is missing or unknown.
 */
enum class SenderClient(val wireName: String) {
    ANDROID("android"),
    WEB("web");

    companion object {
        /** Strict allowlist lookup; null for missing, mistyped, or unknown values. */
        fun fromWire(value: String?): SenderClient? = entries.firstOrNull { it.wireName == value }
    }

    /** Advisory text shown next to a received message on this device. */
    fun originNotice(): String = when (this) {
        WEB -> "Sent from the web client: that device may lack screenshot protection, and its browser cannot guarantee memory wiping."
        ANDROID -> "Sent from the Android app: the sending device enforces screen-capture and memory protections."
    }
}
